lib/checks/tables/td-headers-attr-evaluate.js

```diff
@@ -1,63 +1,79 @@
 import { tokenList } from '../../core/utils';
 import { isVisibleToScreenReaders } from '../../commons/dom';
+import { getRole } from '../../commons/aria';
+
+// Order determines the priority of reporting
+// Only if 0 of higher issues exists will the next be reported
+const messageKeys = [
+  'cell-header-not-in-table',
+  'cell-header-not-th',
+  'header-refs-self',
+  'empty-hdrs' // incomplete
+];
+const [notInTable, notTh, selfRef, emptyHdrs] = messageKeys;
 
 export default function tdHeadersAttrEvaluate(node) {
   const cells = [];
-  const reviewCells = [];
-  const badCells = [];
-
+  const cellRoleById = {};
   for (let rowIndex = 0; rowIndex < node.rows.length; rowIndex++) {
     const row = node.rows[rowIndex];
 
     for (let cellIndex = 0; cellIndex < row.cells.length; cellIndex++) {
-      cells.push(row.cells[cellIndex]);
+      const cell = row.cells[cellIndex];
+      cells.push(cell);
+
+      // Save header id to set if it's th or td with roles columnheader/rowheader
+      const cellId = cell.getAttribute('id');
+      if (cellId) {
+        cellRoleById[cellId] = getRole(cell);
+      }
     }
   }
 
-  const ids = cells
-    .filter(cell => cell.getAttribute('id'))
-    .map(cell => cell.getAttribute('id'));
-
+  const badCells = {
+    [selfRef]: [],
+    [notInTable]: [],
+    [notTh]: [],
+    [emptyHdrs]: []
+  };
   cells.forEach(cell => {
-    let isSelf = false;
-    let notOfTable = false;
-
     if (!cell.hasAttribute('headers') || !isVisibleToScreenReaders(cell)) {
       return;
     }
-
     const headersAttr = cell.getAttribute('headers').trim();
     if (!headersAttr) {
-      return reviewCells.push(cell);
+      badCells[emptyHdrs].push(cell);
+      return;
     }
 
+    const cellId = cell.getAttribute('id');
     // Get a list all the values of the headers attribute
     const headers = tokenList(headersAttr);
-
-    if (headers.length !== 0) {
-      // Check if the cell's id is in this list
-      if (cell.getAttribute('id')) {
-        isSelf = headers.indexOf(cell.getAttribute('id').trim()) !== -1;
+    headers.forEach(headerId => {
```

Comment:
I think this is going to result in an issue where `cell` will be included multiple times in `relatedNodes` if its `headers` references multiple IDs with the same type of issue, ie

```ts
    fixtureSetup(`
      <table>
        <tr> <td id="hi1">hello</td> </tr>
        <tr> <td id="hi2">hello</td> </tr>
        <tr> <td id="bye" headers="hi1 hi2">goodbye</td> </tr>
      </table>'
    `);

    var node = fixture.querySelector('table');

    // assertion fails, _relatedNodes has length 2
    assert.deepEqual(checkContext._relatedNodes, [fixture.querySelector('#bye')]);
```

It'd be nice to improve the tests to exercise `relatedNodes` and to broaden the coverage of multiple-id cases such that they would catch this

Reply:
I reviewed this and came to the conclusion that the node deduplication logic should be located in [check-helper.js](https://github.com/engineerklimov/axe-core/blob/develop/lib/core/utils/check-helper.js#L28). I believe it’s better to keep the check functions slim, and this approach will also help avoid code duplication.

Changing a core utility function, however, seems out of scope for this request, especially considering that this function is used in multiple places, which increases the risk of potential bugs. Additionally, if I were to add code to handle duplicated nodes here, and later someone enhances the relatedNodes function to handle it more generally, there’s a high chance that this logic would end up duplicated in both specific checks and the general utility.

In summary, I recommend handling duplicates directly in relatedNodes and making that enhancement in a separate request.

Reply:
I see where you're coming from, but I strongly agree that it'd be out of scope for this PR; besides being a core utility function, it's part of the published API for custom checks, so at the very least if we made such a change we'd want it to be its own changelog line.

I don't think we should allow this PR to regress the behavior in the meantime, though. I think a straightforward option would be to have `badCells` store `Set`s instead of arrays of cells and normalize it back to an array at the end with `this.relatedNodes([...badCells[messageKey]]);` (we've already established precedent of being willing to use `Set` in #3423)

Reply:
:dart: Set looks a good compromise, added your suggestion.